File: frontend/employee/src/utils/businessDays.ts
export interface HongKongHoliday {
  date: string;
  name: string;
}

let holidaysCache: HongKongHoliday[] = [];
let lastFetch: number = 0;
const CACHE_DURATION = 24 * 60 * 60 * 1000; // 24 hours

export async function fetchHongKongHolidays(): Promise<HongKongHoliday[]> {
  const now = Date.now();
  
  // Return cached data if it's still fresh
  if (holidaysCache.length > 0 && (now - lastFetch) < CACHE_DURATION) {
    return holidaysCache;
  }
  
  try {
    const response = await fetch('https://www.1823.gov.hk/common/ical/en.json');
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    const data = await response.json();
    
    const holidays: HongKongHoliday[] = [];
    
    if (data.vcalendar && data.vcalendar[0] && data.vcalendar[0].vevent) {
      const events = data.vcalendar[0].vevent;
      
      events.forEach((event: any) => {
        if (event.dtstart && event.summary) {
          const dateStr = event.dtstart[0];
          // Convert YYYYMMDD format to YYYY-MM-DD
          const formattedDate = `${dateStr.slice(0, 4)}-${dateStr.slice(4, 6)}-${dateStr.slice(6, 8)}`;
          
          holidays.push({
            date: formattedDate,
            name: event.summary
          });
        }
      });
    }
    
    holidaysCache = holidays;
    lastFetch = now;
    
    return holidays;
  } catch (error) {
    console.error('Failed to fetch Hong Kong holidays:', error);
    
    // Fallback to some basic holidays if fetch fails
    const currentYear = new Date().getFullYear();
    return [
      { date: `${currentYear}-01-01`, name: 'The first day of January' },
      { date: `${currentYear}-05-01`, name: 'Labour Day' },
      { date: `${currentYear}-07-01`, name: 'Hong Kong Special Administrative Region Establishment Day' },
      { date: `${currentYear}-10-01`, name: 'National Day' },
      { date: `${currentYear}-12-25`, name: 'Christmas Day' },
      { date: `${currentYear}-12-26`, name: 'The first weekday after Christmas Day' }
    ];
  }
}

export function isWeekend(date: Date): boolean {
  const day = date.getDay();
  return day === 0 || day === 6; // Sunday = 0, Saturday = 6
}

export function isHongKongPublicHoliday(date: Date, holidays: HongKongHoliday[]): boolean {
  const dateString = date.toISOString().split('T')[0];
  return holidays.some(holiday => holiday.date === dateString);
}

export function isBusinessDay(date: Date, holidays: HongKongHoliday[]): boolean {
  return !isWeekend(date) && !isHongKongPublicHoliday(date, holidays);
}

export async function calculateBusinessDays(startDate: string, endDate: string): Promise<number> {
  const start = new Date(startDate);
  const end = new Date(endDate);
  
  if (start > end) {
    return 0;
  }
  
  const holidays = await fetchHongKongHolidays();
  let businessDays = 0;
  const currentDate = new Date(start);
  
  while (currentDate <= end) {
    if (isBusinessDay(currentDate, holidays)) {
      businessDays++;
    }
    currentDate.setDate(currentDate.getDate() + 1);
  }
  
  return businessDays;
}

export async function getHongKongHolidaysInRange(startDate: string, endDate: string): Promise<HongKongHoliday[]> {
  const start = new Date(startDate);
  const end = new Date(endDate);
  const holidays = await fetchHongKongHolidays();
  
  return holidays.filter(holiday => {
    const holidayDate = new Date(holiday.date);
    return holidayDate >= start && holidayDate <= end;
  });
}

export interface ExcludedDate {
  date: string;
  reason: 'Weekend' | 'Holiday';
  name?: string;
}

export async function getExcludedDatesInRange(startDate: string, endDate: string): Promise<ExcludedDate[]> {
  const start = new Date(startDate);
  const end = new Date(endDate);
  const holidays = await fetchHongKongHolidays();
  const excludedDates: ExcludedDate[] = [];
  
  const currentDate = new Date(start);
  
  while (currentDate <= end) {
    if (isWeekend(currentDate)) {
      excludedDates.push({
        date: currentDate.toISOString().split('T')[0],
        reason: 'Weekend'
      });
    } else if (isHongKongPublicHoliday(currentDate, holidays)) {
      const holiday = holidays.find(h => h.date === currentDate.toISOString().split('T')[0]);
      excludedDates.push({
        date: currentDate.toISOString().split('T')[0],
        reason: 'Holiday',
        name: holiday?.name
      });
    }
    
    currentDate.setDate(currentDate.getDate() + 1);
  }
  
  return excludedDates;
}

File: backend/employee/vacation/business_days.py
import requests
from datetime import datetime, timedelta
from typing import List, Dict
import logging

logger = logging.getLogger(__name__)

class BusinessDaysCalculator:
    _holidays_cache = []
    _last_fetch = None
    _cache_duration = timedelta(hours=24)
    
    @classmethod
    def fetch_hong_kong_holidays(cls) -> List[Dict[str, str]]:
        """Fetch Hong Kong public holidays from government API with caching"""
        now = datetime.now()
        
        # Return cached data if it's still fresh
        if (cls._holidays_cache and cls._last_fetch and 
            (now - cls._last_fetch) < cls._cache_duration):
            return cls._holidays_cache
        
        try:
            response = requests.get('https://www.1823.gov.hk/common/ical/en.json', timeout=10)
            response.raise_for_status()
            
            data = response.json()
            holidays = []
            
            if 'vcalendar' in data and data['vcalendar'] and 'vevent' in data['vcalendar'][0]:
                events = data['vcalendar'][0]['vevent']
                
                for event in events:
                    if 'dtstart' in event and 'summary' in event:
                        date_str = event['dtstart'][0]
                        # Convert YYYYMMDD format to YYYY-MM-DD
                        formatted_date = f"{date_str[:4]}-{date_str[4:6]}-{date_str[6:8]}"
                        
                        holidays.append({
                            'date': formatted_date,
                            'name': event['summary']
                        })
            
            cls._holidays_cache = holidays
            cls._last_fetch = now
            
            return holidays
            
        except Exception as e:
            logger.error(f"Failed to fetch Hong Kong holidays: {e}")
            
            # Fallback to some basic holidays if fetch fails
            current_year = now.year
            return [
                {'date': f'{current_year}-01-01', 'name': 'The first day of January'},
                {'date': f'{current_year}-05-01', 'name': 'Labour Day'},
                {'date': f'{current_year}-07-01', 'name': 'Hong Kong Special Administrative Region Establishment Day'},
                {'date': f'{current_year}-10-01', 'name': 'National Day'},
                {'date': f'{current_year}-12-25', 'name': 'Christmas Day'},
                {'date': f'{current_year}-12-26', 'name': 'The first weekday after Christmas Day'}
            ]
    
    @classmethod
    def is_weekend(cls, date: datetime) -> bool:
        """Check if date is a weekend (Saturday=5, Sunday=6)"""
        return date.weekday() >= 5
    
    @classmethod
    def is_hong_kong_public_holiday(cls, date: datetime, holidays: List[Dict[str, str]]) -> bool:
        """Check if date is a Hong Kong public holiday"""
        date_str = date.strftime('%Y-%m-%d')
        return any(holiday['date'] == date_str for holiday in holidays)
    
    @classmethod
    def is_business_day(cls, date: datetime, holidays: List[Dict[str, str]]) -> bool:
        """Check if date is a business day (not weekend and not public holiday)"""
        return not cls.is_weekend(date) and not cls.is_hong_kong_public_holiday(date, holidays)
    
    @classmethod
    def calculate_business_days(cls, start_date: datetime, end_date: datetime) -> int:
        """Calculate number of business days between two dates (inclusive)"""
        if start_date > end_date:
            return 0
        
        holidays = cls.fetch_hong_kong_holidays()
        business_days = 0
        current_date = start_date
        
        while current_date <= end_date:
            if cls.is_business_day(current_date, holidays):
                business_days += 1
            current_date += timedelta(days=1)
        
        return business_days
    
    @classmethod
    def get_holidays_in_range(cls, start_date: datetime, end_date: datetime) -> List[Dict[str, str]]:
        """Get all Hong Kong holidays within the specified date range"""
        holidays = cls.fetch_hong_kong_holidays()
        
        return [
            holiday for holiday in holidays
            if start_date <= datetime.strptime(holiday['date'], '%Y-%m-%d').date() <= end_date
        ]